src/nsrt_learning/sampler_learning.py

```diff
@@ -245,22 +275,21 @@ def _create_sampler_data(
             var_types = [var.type for var in variables]
             objects = list(state)
             for grounding in utils.get_object_combinations(objects, var_types):
+                if len(negative_data
+                       ) >= CFG.sampler_learning_max_negative_data:
+                    # If we already have more negative examples
+                    # than the maximum specified in the config,
+                    # we don't add any more negative examples.
+                    return positive_data, negative_data
+
                 # If we are currently at the datastore that we're learning a
-                # sampler for, and this datapoint matches the actual grounding,
-                # add it to the positive data and continue.
+                # sampler for, and this datapoint matches the positive
+                # grounding, this was already added to the positive data, so
+                # we can continue.
                 if idx == datastore_idx:
-                    actual_grounding = [var_to_obj[var] for var in variables]
-                    if grounding == actual_grounding:
-                        assert all(
-                            pre.predicate.holds(
-                                state, [var_to_obj[v] for v in pre.variables])
-                            for pre in preconditions)
-                        positive_data.append((state, var_to_obj, option, goal))
+                    positive_grounding = [var_to_obj[var] for var in variables]
```

Comment:
move this outside the for loop (one level)?

Reply:
good catch!

Reply:
note: actually this is wrong, reverting the change. this line only makes sense if `idx == datastore_idx` so it needs to be here

Reply:
I don't get it but okay 🙃

Reply:
same, let's not ask too many questions

Reply:
the real explanation is that `variables` is only associated with `datastores[datastore_idx]`, so other datastores may have entirely different variables, and so this line doesn't make sense for them